### PR TITLE
Use padraic/phar-updater to handle PAHR updates

### PR DESCRIPTION
**Pull request by jubianchi**
It's been a while since I sent my last contribution. So here is a new one!

It adds a new `self-update` command to pickle letting users update their local Phar when a new release is issued on github with a simple command line: `sudo pickle self-update`

It has currently one option `--unstable` to update to unstable release.
I plan to add a `--rollback` option to let the users revert the update if anything went wrong.

Let me know what you think about it ;)

### CODE REVIEW

File: src/Console/Command/SelfUpdateCommand.php [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+                null,
+                InputOption::VALUE_NONE,
+                'Update to a non-stable version'
+            )
+        ;
+    }
+
+    protected function execute(InputInterface $input, OutputInterface $output)
+    {
+        if (\Phar::running() === '') {
```
Comment (stof) [inline src/Console/Command/SelfUpdateCommand.php]: You should rather disable the command entirely outside a phar

Comment (jubianchi) [inline src/Console/Command/SelfUpdateCommand.php]: true

File: src/Console/Command/SelfUpdateCommand.php [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+            throw new \RuntimeException($this->getName() . ' can only be run from the Phar!');
+        }
+
+        $updater = new Updater(null, false, Updater::STRATEGY_GITHUB);
+        $updater->getStrategy()->setPackageName('friendsofphp/pickle');
+        $updater->getStrategy()->setPharName(self::PHAR_NAME);
+        $updater->getStrategy()->setCurrentLocalVersion('v0.5.0');
```
Comment (padraic) [inline src/Console/Command/SelfUpdateCommand.php]: It's might be worth moving this somewhere more obvious. Don't forget to update this when tagging... Any way of automating it using the PHAR build process?

File: src/Console/Command/SelfUpdateCommand.php [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+        if ($input->getOption('unstable')) {
+            $updater->getStrategy()->setStability('unstable');
+        }
+
+        if ($updater->update() === false) {
```
Comment (padraic) [inline src/Console/Command/SelfUpdateCommand.php]: See https://github.com/padraic/humbug/blob/master/src/Command/SelfUpdate.php and README

The update() method returns a boolean. True for updated, and False where no update was available. It throws an Exception on other errors, would typically be signature or HTTP errors. So you should put that line in a `try...catch` block and use the result to tailor the message for either a successful update or no update available, reporting the error in the `catch` as needed.

**Comment by shouze** [comment]
@jubianchi how can I help to make this one mergeable & a new release tag available ?
This will be a good move for pickle adoption. Also, php-build [has recently dropped pear & pyrus](https://github.com/php-build/php-build/pull/347) so for example pickle is the best candidate to replace them right now on travis-ci for example.

I had a discussion yesterday with @BanzaiMan https://github.com/travis-ci/travis-ci/issues/5256#issuecomment-163184941

File: src/Console/Command/SelfUpdateCommand.php (hunk 2 above)
Comment (shouze) [inline src/Console/Command/SelfUpdateCommand.php]: As suggested by @stof https://github.com/travis-ci/travis-ci/issues/5266#issuecomment-164808764 using github api to know the latest release looks like to be a good thing no? It depends if rolling release is the choice or not... composer self-update way is not so bad in practice.

Comment (stof) [inline src/Console/Command/SelfUpdateCommand.php]: well, storing the phar in github releases requires doing a github release, which means tagging a release. this does not go the rolling-release way.
Composer stores phars on their own server (btw, they support updating to releases too).

anyway, this string is not the version of the latest release. It is the version of the current code (so it should match the tag for which it gets built)

Comment (Hywan) [inline src/Console/Command/SelfUpdateCommand.php]: @shouze, @stof, @padraic: We are not doing rolling-release and a PHAR is attached to each release with **the same name**, so looking at Github API is correct.

**Comment by pierrejoye** [comment]
Any hand need here or is someone working on merging it? :)

**Comment by jubianchi** [comment]
everything should be OK now ;)

**Comment by pierrejoye** [comment]
Expected to have travis failed?
If yes go ahead and merge it, thanks a lot :)

File: src/Console/Application.php [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+            "json",
+            "dom",
+            "openssl",
+            "phar",
+            "zip",
+        );
+
+        foreach ($required_exts as $ext) {
+            if (!extension_loaded($ext)) {
+                die("Extension '$ext' required but not loaded, full required list: " . implode(", ", $required_exts));
```
Comment (stof) [inline src/Console/Application.php]: you should actually write this to stderr and die with a failure exit code instead. This will use a success exit code

**Pull request merged**